File: Decorator/DecoratorPattern.py
# coding: utf8

import abc


class Component(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def process(self):
        pass


class ConcreteComponent(Component):
    def process(self):
        print("process in ConcreteComponent")


class Decorator(Component):
    def __init__(self, component):
        self._component = component

    def process(self):
        self.before()
        self._component.process()
        self.after()

    @abc.abstractmethod
    def before(self):
        pass

    @abc.abstractmethod
    def after(self):
        pass


class LogDecorator(Decorator):
    def before(self):
        print("before calling process")

    def after(self):
        print("after calling process")


if __name__ == "__main__":
    import unittest


    class DecoratorTest(unittest.TestCase):
        def testDecorator(self):
            component = ConcreteComponent()
            decorator = LogDecorator(component)
            decorator.process()

    unittest.main()

File: Decorator/DecoratorPattern.java
interface Component {
    void operation();
}

class ConcreteComponent implements Component {
    public void operation() {
        System.out.println("operation in ConcreteComponent");
    }
}

abstract class LogDecrator {
    private Component component;

    public LogDecrator(Component component) {
        if (component == null) 
            throw new RuntimeException("component == null");
        this.component = component;
    }

    public void operation() {
        beforeOperation();
        component.operation();
        afterOperation();
    }

    public abstract void beforeOperation();
    public abstract void afterOperation();
}

class ConreteLogDecorator extends LogDecrator {
    public ConreteLogDecorator(Component component) {
        super(component);
    }

    public void beforeOperation() {
        System.out.println("before operation");
    }

    public void afterOperation() {
        System.out.println("after operation");
    }
}

class DecoratorPattern {
    public static void main(String[] args) {
        Component component =  new ConcreteComponent();
        LogDecrator logDecorator = new ConreteLogDecorator(component);
        logDecorator.operation();
    }
}

